File: backend/services/api-document/src/application/upload_file.py

import asyncio
import json
import time
from uuid import UUID
from uuid6 import uuid7
from asyncpg.exceptions import ForeignKeyViolationError, UniqueViolationError
from fastapi import File, Form, HTTPException, Request, status, UploadFile
from loguru._logger import Logger
from mypy_boto3_s3 import S3Client
from pydantic import BaseModel
from .router import router

# Map of supported MIME types to their corresponding file extensions
ALLOWED_MIME_TYPES: dict[str, str] = {
    "application/pdf": "pdf",
    "application/vnd.openxmlformats-officedocument.wordprocessingml.document": "docx",
    "text/markdown": "md",
    "text/plain": "txt",
}
# Maximum file size limit: 50 MB
MAX_FILE_SIZE_BYTES = 100 * 1024 * 1024 

async def validate_file(file: UploadFile) -> tuple[bytes, str, str]:
    """
    Returns:
        (contents, mime_type, extension)
    Raises:
        HTTPException 415 — unsupported type
        HTTPException 400 — empty file
        HTTPException 413 — exceeds size limit
    """
    # 1. Validate MIME type against the allowed list
    if file.content_type not in ALLOWED_MIME_TYPES:
        raise HTTPException(
            status_code=status.HTTP_415_UNSUPPORTED_MEDIA_TYPE,
            detail=(
                f"Unsupported file type: '{file.content_type}'. "
                f"Allowed types: {', '.join(ALLOWED_MIME_TYPES.keys())}"
            ),
        )
        
    # 2. Read file content into memory
    contents = await file.read()
    
    # 3. Ensure the file is not empty
    if not contents:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="The uploaded file is empty.",
        )
        
    # 4. Validate file size
    if len(contents) > MAX_FILE_SIZE_BYTES:
        raise HTTPException(
            status_code=status.HTTP_413_REQUEST_ENTITY_TOO_LARGE,
            detail=f"File size exceeds the {MAX_FILE_SIZE_BYTES // (1024 * 1024)} MB limit.",
        )
        
    # 5. Return validated data, MIME type, and the mapped extension
    return contents, file.content_type, ALLOWED_MIME_TYPES[file.content_type]

async def delete_s3_object(s3: S3Client, key: str, bucket: str, logger: Logger) -> None:
    try:
        loop = asyncio.get_running_loop()
        await loop.run_in_executor(
            None,
            lambda: s3.delete_object(Bucket=bucket, Key=key),
        )     
        
        logger.info(f"[S3Service] Compensating DELETE succeeded: {key}")
    except Exception as e:
        logger.error(f"[S3Service] Compensating DELETE failed for '{key}': {e}")
    
  
# Response schema
class DocumentResponse(BaseModel):
    id: UUID
    folder_id: UUID
    original_name: str
    internal_name: str
    content_type: str
    mime_type: str
    size_bytes: int
    storage_path: str      
    created_at: int

def _build_metadata(original_name: str, size_bytes: int, folder_id: str) -> str:
    """
    JSONB metadata stored alongside the document record.
    Extend freely — this column is the right place for future
    fields (page count, language, virus-scan result, etc.)
    without requiring schema migrations.
    """
    return json.dumps({
        "original_name": original_name,
        "size_bytes":    size_bytes,
        "folder_id":     folder_id,
    })


# ── Endpoint ──────────────────────────────────────────────────

@router.post(
    "/upload-file",
    status_code=status.HTTP_201_CREATED,
    response_model=DocumentResponse,
)
async def upload_file(
    request: Request,
    file: UploadFile = File(...),
    folder_id: str = Form(...)
):
    """
    Rollback flow:
      1. Upload to SeaweedFS  →  if it fails, DB is untouched.
      2. INSERT into DB       →  if it fails, compensate by deleting from S3.
    """
    logger = request.app.state.logger
    s3 = request.app.state.s3
    pool = request.app.state.pool  
    bucket = request.app.state.settings.s3_bucket

    logger.info(f"filename={file.filename!r}")
    logger.info(f"content_type={file.content_type!r}")
    logger.info(f"folder_id={folder_id!r}")
        
    contents, raw_content_type, ext = await validate_file(file)

    user_id       = "019d2612-a01d-734c-ab63-917106f31187"  # TODO: authentication
    doc_id        = uuid7()
    original_name = file.filename or f"document.{ext}"
    internal_name = f"{doc_id}.{ext}"
    storage_key   = f"{user_id}/{folder_id}/{internal_name}"
    created_at    = int(time.time() * 1000)
    size_bytes   = len(contents)
    metadata      = _build_metadata(original_name, size_bytes, folder_id)
    initial_v = 0
    # content_type  → raw HTTP header value, may include params
    #                 e.g. "application/pdf; name=report.pdf"
    # mime_type     → clean MIME type, no params
    #                 e.g. "application/pdf"
    content_type = raw_content_type
    mime_type    = raw_content_type.split(";")[0].strip()
    
    
    SQL_INSERT = """
    INSERT INTO documents (
        id, user_id, folder_id,
        original_name, internal_name,
        content_type, mime_type,
        size_bytes, storage_path,
        metadata,
        created_at, readed_at, updated_at, deleted_at,
        v
    ) VALUES (
        $1,  $2,  $3,
        $4,  $5,
        $6,  $7,
        $8,  $9,
        $10,
        $11, NULL, NULL, NULL,
        $12
    )
    RETURNING *;
    """

    # ── Phase 1: upload binary ────────────────────────────────
    try:
        loop = asyncio.get_running_loop()
        await loop.run_in_executor(
            None,
            lambda: s3.put_object(
                Bucket=bucket,
                Key=storage_key,
                Body=contents,
                ContentType=mime_type,
                Metadata={
                "document-id":   str(doc_id),
                "folder-id":     folder_id,
                "user-id":       user_id,
                "original-name": original_name,
                "internal-name": internal_name,
                }
            ),
        )
                
        logger.info(f"[upload-document] S3 object written: {storage_key}")
    except Exception as e:
        logger.error(f"[upload-document] SeaweedFS upload failed: {e}")
        raise HTTPException(status.HTTP_502_BAD_GATEWAY, "Error uploading file to S3")

    # ── Phase 2: persist metadata ─────────────────────────────
    async with pool.acquire() as connection:
        try:
            async with connection.transaction():
                row = await connection.fetchrow(
                    SQL_INSERT,
                    doc_id,        # $1  id
                    user_id,       # $2  user_id
                    folder_id,     # $3  folder_id
                    original_name, # $4  original_name
                    internal_name, # $5  internal_name
                    content_type,  # $6  content_type
                    mime_type,     # $7  mime_type
                    size_bytes,    # $8  size_bytes
                    storage_key,   # $9  storage_path
                    metadata,      # $10 metadata
                    
                    created_at,    # $11 created_at
                    initial_v      # $12 v
                )
                if not row:
                    raise RuntimeError("INSERT RETURNING returned empty.")

                data_json = json.dumps(dict(row), default=str)
                metadata = { "user_id": user_id }
                meta_json = json.dumps(metadata or {}, default=str)
                    
                _EVENT_QUERY = """
                INSERT INTO events (
                    specversion, event_type, source, id, time, 
                    entity_type, entity_id, data, metadata
                ) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9)
                """
                
                await connection.execute(
                    _EVENT_QUERY,
                    0, 
                    'document.created',             
                    'api-document', 
                    uuid7(),                   
                    int(time.time() * 1000),        
                    'document',              
                    str(row["id"]),                
                    data_json,                  
                    meta_json                 
                ) 
                
        except ForeignKeyViolationError:
            await delete_s3_object(s3, storage_key, bucket, logger)
            raise HTTPException(status.HTTP_404_NOT_FOUND, f"The folder '{folder_id}' does not exist.")

        except UniqueViolationError:
            await delete_s3_object(s3, storage_key, bucket, logger)
            raise HTTPException(status.HTTP_409_CONFLICT, "A document with that identifier already exists.")

        except HTTPException:
            raise

        except Exception as e:
            logger.error(f"[upload-document] DB insert failed: {e}")
            await delete_s3_object(s3, storage_key, bucket, logger)
            raise HTTPException(status.HTTP_500_INTERNAL_SERVER_ERROR, "Error registering the document.")

    # TODO: Dispatch Pub/Sub event

    return DocumentResponse(**dict(row))

File: backend/services/api-document/src/application/upload-file.ts
import {
  PutObjectCommand,
  DeleteObjectCommand,
  S3Client,
} from "@aws-sdk/client-s3";
import { FastifyRequest, FastifyReply } from "fastify";
import { v7 as uuidv7 } from "uuid";
import type { File } from "fastify-multer/lib/interfaces";
import { pool } from "../infrastructure/postgres/db.js";
import multer from "fastify-multer";
import z from "zod";

/**
 * Map of supported MIME types to their corresponding file extensions.
 */
const ALLOWED_MIME_TYPES: Record<string, string> = {
  "application/pdf": "pdf",
  "application/vnd.openxmlformats-officedocument.wordprocessingml.document":
    "docx",
  "text/markdown": "md",
  "text/plain": "txt",
};

/**
 * Maximum allowed file size (100 MB).
 */
const MAX_FILE_SIZE_BYTES = 100 * 1024 * 1024;

/**
 * Multer middleware configuration for handling multipart/form-data.
 * Uses memory storage and validates file types against ALLOWED_MIME_TYPES.
 */
export const uploadMiddleware = multer({
  storage: multer.memoryStorage(),
  limits: { fileSize: MAX_FILE_SIZE_BYTES },
  fileFilter: (_req, file, cb) => {
    const mime = file.mimetype.split(";")[0].trim();
    if (mime in ALLOWED_MIME_TYPES) {
      cb(null, true);
    } else {
      cb(
        new Error(
          `Unsupported file type: '${mime}'. Allowed types: ${Object.keys(ALLOWED_MIME_TYPES).join(", ")}`,
        ),
      );
    }
  },
});

/**
 * Deletes an object from S3. Typically used as a compensating action
 * if a subsequent database transaction fails.
 * @param s3 - The S3 client instance.
 * @param bucket - S3 bucket name.
 * @param key - The object key (path) to delete.
 * @param log - Fastify logger instance.
 */
async function deleteS3Object(
  s3: S3Client,
  bucket: string,
  key: string,
  log: FastifyRequest["log"],
): Promise<void> {
  try {
    await s3.send(new DeleteObjectCommand({ Bucket: bucket, Key: key }));
    log.info(`[S3Service] Compensating DELETE succeeded: ${key}`);
  } catch (e) {
    log.error(`[S3Service] Compensating DELETE failed for '${key}': ${e}`);
  }
}

/**
 * Constructs a JSON string containing file metadata.
 * @param originalName - The original name of the uploaded file.
 * @param sizeBytes - File size in bytes.
 * @param folderId - UUID of the parent folder.
 * @returns A JSON stringified metadata object.
 */
function buildMetadata(
  originalName: string,
  sizeBytes: number,
  folderId: string,
): string {
  return JSON.stringify({
    original_name: originalName,
    size_bytes: sizeBytes,
    folder_id: folderId,
  });
}

// ── SQL ────────────────────────────────────────────────────────

const SQL_CHECK_FOLDER_OWNERSHIP = `
  SELECT 1 FROM folders WHERE id = $1 AND user_id = $2;
`;

const SQL_CHECK_FOLDER_OWNERSHIP_LOCKED = `
  SELECT 1 FROM folders WHERE id = $1 AND user_id = $2 FOR SHARE;
`;

const SQL_INSERT = `
  INSERT INTO documents (
    id, user_id, folder_id,
    original_name, internal_name,
    content_type, mime_type,
    size_bytes, storage_path,
    metadata,
    created_at, readed_at, updated_at, deleted_at,
    v
  ) VALUES (
    $1,  $2,  $3,
    $4,  $5,
    $6,  $7,
    $8,  $9,
    $10,
    $11, NULL, NULL, NULL,
    $12
  )
  RETURNING *;
`;

const SQL_INSERT_EVENT = `
  INSERT INTO events (
    specversion, event_type, source, id, time,
    entity_type, entity_id, data, metadata
  ) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9)
`;

/**
 * Zod schema for validating the upload request body.
 */
export const UploadFileBody = z.object({
  folder_id: z.uuid("folder_id must be a valid UUID"),
});

/**
 * Main handler for uploading files.
 * * Process:
 * 1. Validates the request body and file existence.
 * 2. Uploads the file buffer to AWS S3.
 * 3. Starts a DB transaction to record document metadata and a 'document.created' event.
 * 4. Performs a rollback and deletes the S3 object if the DB transaction fails.
 * @param request - Fastify request object containing the file and body.
 * @param reply - Fastify reply object.
 */
export async function uploadFileHandler(
  request: FastifyRequest,
  reply: FastifyReply,
) {
  const { s3, config, log } = request.server;

  const file = request.file as File;
  if (!file) {
    return reply.status(400).send({ message: "No file provided." });
  }

  const parsed = UploadFileBody.safeParse(request.body);
  if (!parsed.success) {
    const formatted = z.treeifyError(parsed.error);

    return reply.status(400).send({
      error: "Validation error",
      details: formatted,
    });
  }

  const { folder_id: folderId } = parsed.data;

  const userId = "019d2612-a01d-734c-ab63-917106f31187"; // TODO: authentication

  const folderCheck = await pool.query(SQL_CHECK_FOLDER_OWNERSHIP, [
    folderId,
    userId,
  ]);

  if (folderCheck.rowCount === 0) {
    return reply.status(403).send({
      message:
        "Access denied: Folder does not belong to user or does not exist.",
    });
  }

  const fileBuffer = file.buffer!;
  const rawContentType = file.mimetype; // "application/pdf; charset=utf-8"
  const mimeType = rawContentType.split(";")[0].trim(); // "application/pdf"
  const extension = ALLOWED_MIME_TYPES[mimeType];
  const originalName = file.originalname || `document.${extension}`;
  const sizeBytes = file.size!;

  log.debug(`filename=${originalName}`);
  log.debug(`content_type=${mimeType}`);
  log.debug(`folder_id=${folderId}`);

  const docId = uuidv7();
  const internalName = `${docId}.${extension}`;
  const storageKey = `${userId}/${folderId}/${internalName}`;
  const createdAt = Date.now();
  const metadata = buildMetadata(originalName, sizeBytes, folderId);
  const initialV = 0;

  // ── Phase 1: upload binary ────────────────────────────────

  try {
    await s3.send(
      new PutObjectCommand({
        Bucket: config.s3.bucket,
        Key: storageKey,
        Body: fileBuffer,
        ContentType: mimeType,
        Metadata: {
          "document-id": docId,
          "folder-id": folderId,
          "user-id": userId,
          "original-name": originalName,
          "internal-name": internalName,
        },
      }),
    );
    log.info(`[upload-document] S3 object written: ${storageKey}`);
  } catch (e) {
    log.error(`[upload-document] S3 upload failed: ${e}`);
    return reply.status(502).send({ message: "Error uploading file to S3." });
  }

  // ── Phase 2: persist metadata ─────────────────────────────

  const client = await pool.connect();

  try {
    await client.query("BEGIN");

    const folderCheck = await client.query(SQL_CHECK_FOLDER_OWNERSHIP_LOCKED, [
      folderId,
      userId,
    ]);

    if (folderCheck.rowCount === 0) {
      await client.query("ROLLBACK");
      await deleteS3Object(s3, config.s3.bucket, storageKey, log);
      return reply.status(403).send({
        message:
          "Access denied: Folder does not belong to user or does not exist.",
      });
    }

    const { rows } = await client.query(SQL_INSERT, [
      docId, // $1  id
      userId, // $2  user_id
      folderId, // $3  folder_id
      originalName, // $4  original_name
      internalName, // $5  internal_name
      rawContentType, // $6  content_type
      mimeType, // $7  mime_type
      sizeBytes, // $8  size_bytes
      storageKey, // $9  storage_path
      metadata, // $10 metadata
      createdAt, // $11 created_at
      initialV, // $12 v
    ]);

    const document = rows[0];
    if (!document) throw new Error("INSERT RETURNING returned empty.");

    const normalizedDocument = {
      ...document,
      size_bytes: Number(document.size_bytes),
      created_at: Number(document.created_at),
      v: Number(document.v),
    };

    await client.query(SQL_INSERT_EVENT, [
      0, // $1 specversion
      "document.created", // $2 event_type
      "api-document", // $3 source
      uuidv7(), // $4 id
      Date.now(), // $5 time
      "document", // $6 entity_type
      document.id, // $7 entity_id
      normalizedDocument, // $8 data
      { user_id: userId }, // $9 metadata
    ]);

    await client.query("COMMIT");

    // TODO: Dispatch Pub/Sub event

    return reply.status(201).send(normalizedDocument);
  } catch (e: any) {
    await client.query("ROLLBACK");

    if (e.code === "23503") {
      await deleteS3Object(s3, config.s3.bucket, storageKey, log);
      return reply
        .status(404)
        .send({ message: `The folder '${folderId}' does not exist.` });
    }

    if (e.code === "23505") {
      await deleteS3Object(s3, config.s3.bucket, storageKey, log);
      return reply
        .status(409)
        .send({ message: "A document with that identifier already exists." });
    }

    log.error(`[upload-document] DB insert failed: ${e}`);
    await deleteS3Object(s3, config.s3.bucket, storageKey, log);
    return reply
      .status(500)
      .send({ message: "Error registering the document." });
  } finally {
    client.release();
  }
}
